url901cc5721b_shidasan_KonohaCoffee_tgz-pJ8-org_KonohaScript_KonohaJ8/src/org/KonohaScript/UntypedNode.java

```diff
@@ -0,0 +1,370 @@
+		SetAt(Index, Token);
+	}
+	
+	public int ReportError(KToken Token, String Message, int ParseOption) {
+		if((ParseOption & HasNextPattern) == HasNextPattern) {
+			if(Token != null) {
+				KeyToken = Token;
+			}
+			NodeNameSpace.Message(Error, KeyToken, Message);
+			this.Syntax = KSyntax.ErrorSyntax;
+			this.NodeList = null;
+		}
+		return NoMatch;
+	}
+	
+	public static UntypedNode NewNullNode(KNameSpace ns, ArrayList<KToken> TokenList, int BeginIdx) {
+		KToken EmptyToken = new KToken("");
+		EmptyToken.uline = TokenList.get(BeginIdx-1).uline;
+		EmptyToken.ResolvedSyntax = KSyntax.EmptySyntax;
+		return new UntypedNode(ns, EmptyToken);
+	}
+	
+	public int ParseToken(KToken KeyToken, ArrayList<KToken> TokenList, int BeginIdx, int EndIdx) {
+		this.KeyToken = KeyToken;
+		KSyntax Syntax = KeyToken.ResolvedSyntax;
+		int NextIdx = NoMatch;
+		while(Syntax != null) {
+			this.Syntax = Syntax;
+			KonohaDebug.P("(^^;) trying matching.. : " + Syntax.SyntaxName + ":" + Syntax.ParseMethod.getName());
+			NextIdx = Syntax.InvokeParseFunc(this, TokenList, BeginIdx, EndIdx, (Syntax.ParentSyntax == null) ? 0 : HasNextPattern);
+			if(NextIdx != NoMatch) {
+				KonohaDebug.P("(^^;) Matched: " + Syntax.SyntaxName + ":"+Syntax.ParseMethod.getName());
+				return NextIdx;
+			}
+			this.Syntax = null;
+			this.NodeList = null;
+			Syntax = Syntax.ParentSyntax;
+		}
+		ReportError(KeyToken, "undefined syntax: " + KeyToken.ParsedText, 0);
+		return EndIdx;
+	}
+
+	public final static int LeftTerm  = 0;
+	public final static int RightTerm = 1;
+
+	/* 1 + 2 * 3 */
+	/* 1 * 2 + 3 */
+	
+	public static UntypedNode BinaryNode(KNameSpace ns, UntypedNode LeftNode, KToken KeyToken, UntypedNode RightNode)
+	{
+		if(RightNode.Syntax.IsBinaryOperator()) {
+			if(KeyToken.ResolvedSyntax.IsLeftJoin(RightNode.Syntax)) {
+				UntypedNode NewNode = new UntypedNode(ns, KeyToken);
+				NewNode.SetAtNode(LeftTerm, LeftNode);
+				NewNode.SetAtNode(RightTerm, (UntypedNode)(RightNode.NodeList.get(LeftTerm)));
+				RightNode.NodeList.set(LeftTerm, NewNode);
+				return RightNode;
+			}
+		}
+		UntypedNode NewNode = new UntypedNode(ns, KeyToken);
+		NewNode.SetAtNode(LeftTerm, LeftNode);
+		NewNode.SetAtNode(RightTerm, RightNode);
+		if(RightNode.NextNode != null) {
+			NewNode.LinkNode(RightNode.NextNode);
+			RightNode.NextNode = null;
+		}		
+		return NewNode;
+	}
+	
+	int ReportExpectedAfter(ArrayList<KToken> TokenList, int BeginIdx, int EndIdx, String NextSyntax, int ParseOption) {
+		if(BeginIdx == NoMatch) return NoMatch;
+		if((ParseOption & AllowSkip) == AllowSkip) return BeginIdx;
+		if((ParseOption & AllowEmpty) != AllowEmpty) {
+			KToken Token = TokenList.get(BeginIdx-1);
+			return ReportError(Token, NextSyntax + " is expected after " + Token.ParsedText, ParseOption);
+		}
+		return BeginIdx;
+	}
+	
+	public static UntypedNode ParseNewNode2(KNameSpace ns, UntypedNode PrevNode, ArrayList<KToken> TokenList, int BeginIdx, int EndIdx, int ParseOption) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.